### PR TITLE
[Review Only] APEXMALHAR-1701: Deduper using Windowed Operator - Initial Commit. Ad…

### PR DESCRIPTION
**Pull request by bhupeshchawda**
…ded Tests

Added an implementation based on Windowed Operator

### CODE REVIEW

File: library/src/test/java/org/apache/apex/malhar/lib/dedup/WindowedDeduperTest.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  public static void setup()
+  {
+    dedup = new WindowedDeduperImpl();
+    // Define accumulation
+    Accumulation dedupAccum = new DedupAccumulation<TestPojo>();
+    dedup.setAccumulation(dedupAccum);
+    // Set storage to InMemory implementation
+    dedup.setDataStorage(new InMemoryWindowedStorage<List<TestPojo>>());
+    // Set retraction storage. Not used, but is required by the abstract operator.
+    dedup.setRetractionStorage(new InMemoryWindowedStorage<TestPojo>());
```
Comment (davidyan74) [inline library/src/test/java/org/apache/apex/malhar/lib/dedup/WindowedDeduperTest.java:76]: Should not need this if you're not doing triggers at all here. Does it give you an error?

File: library/src/main/java/org/apache/apex/malhar/lib/dedup/AbstractWindowedDeduper.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+   */
+  @Override
+  public void accumulateTuple(Tuple.WindowedTuple<T> tuple)
+  {
+    for (Window window : tuple.getWindows()) { // There will be exactly one window
+      // process each window
+      List<T> accum = dataStorage.get(window);
+      if (accum == null) {
+        // New window?
+        accum = new ArrayList<T>();
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/dedup/AbstractWindowedDeduper.java:81]: I think it's better to call defaultAccumulatedValue() here and have defaultAccumulatedValue return a new list.

File: library/src/test/java/org/apache/apex/malhar/lib/dedup/WindowedDeduperTest.java (hunk 1 above)
Comment (bhupeshchawda) [inline library/src/test/java/org/apache/apex/malhar/lib/dedup/WindowedDeduperTest.java:76]: Yes. It throws a NPE. Changes are needed in the abstract operator.

File: library/src/main/java/org/apache/apex/malhar/lib/dedup/AbstractWindowedDeduper.java (hunk 2 above)
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/dedup/AbstractWindowedDeduper.java:81]: agreed. Will change this.

File: library/src/test/java/org/apache/apex/malhar/lib/dedup/WindowedDeduperTest.java (hunk 1 above)
Comment (davidyan74) [inline library/src/test/java/org/apache/apex/malhar/lib/dedup/WindowedDeduperTest.java:76]: I see where the problem is. I have a fix here in my newest review-only PR: #345

**Comment by bhupeshchawda** [comment]
@chinmaykolhatkar @chandnisingh @sandeepdeshmukh Please review

Please also compare this with https://github.com/apache/apex-malhar/pull/335
We need to decide one implementation to go ahead with.
Thanks!

**Comment by bhupeshchawda** [comment]
Closing the PR. This was some exploratory implementation to see how useful a windowed deduper would be. Here is the new JIRA: https://issues.apache.org/jira/browse/APEXMALHAR-2177

https://issues.apache.org/jira/browse/APEXMALHAR-1701 will be closed as deduper based on managed state is already merged.

**Pull request closed without merging**